### PR TITLE
compaction: introduce multilevel write-amp heuristic

### PR DESCRIPTION
**Pull request by msbutler**
This patch introduces the multiLevelHueristic write amp heuristic.

### CODE REVIEW

**Comment by msbutler** [comment]
@sumeerbhola @itsbilal I've got some interesting benchmarking results to report!

**Setup**
- Run ycsb F until LSM is 300 GB large, varying number of workers, on an n2-standard-16 with a 2TB pd-ssd. See [script](https://github.com/msbutler/crl-scripts/blob/fc8781a0444a0e83cc247acf6bf5569516fe72f0/pebble_multi/pebble.sh). 
- Treatment [sha](https://github.com/cockroachdb/pebble/pull/2179/commits/0415eeb51face21ba51adb5c6dd1c81dd118d61b) is the latest on this PR and control [sha](https://github.com/cockroachdb/pebble/commit/0893071d8a52aa92bc58ef2e1e08e801edb11b87) is the latest commit merged to master rebased on this PR

**Results**
image
- See the LSM metric summary for each trial [here](https://chill-stick-20a.notion.site/trials-6c74e146d3874d62a8276de97d74d45a). 

Initial thoughts:
- At a high level, this heuristic does not seem to affect perf in a healthy LSM (see 64 worker results), and can prevent and mitigate the negative consequences  of an inverted LSM (see 96 and 128 worker results).
- According to gce observability metrics, write throughput declined overtime for the 96 and 128 worker trials for both treatment and control. Bilal and I confirmed that this is likely due to L0 refusing to accept some writes periodically ( the r-amp for L0 is 999 in control-128). Given the superior perf results in the treatment arm, this likely happened less frequently in the treatment arm. 
- I'm unsure if the 300 GB LSM represents a mature LSM, for 64 workers, here's the count/size by level. WDYT?
image


**Next steps**
- Run the 64 worker workload until 600 GB is reached, to see how the maturity of the LSM affects things. I don't think I need to run the 96 and 128 worker configs, as it seems pretty clear that this heuristic helps out a lot.
- Anything else you want me to try?

**Comment by sumeerbhola** [comment]
Thanks for the results. Does look very promising. I haven't yet looked at the code changes.

@jbowens are we at the point where we can run with the compaction benchmarking tooling? I'd be curious what we see in write amp if we were not letting compactions (and L0) fall so far behind.

**Comment by msbutler** [comment]
@sumeerbhola reviewable does not seem to be working, but to answer your question on multilevel compaction count:

By number of workers:
- 64: 12k
- 96:  17k
- 128: 13k

**Review by sumeerbhola — COMMENTED** [review]
I added some byte metrics and am running the KV0 experiment mentioned in https://cockroachlabs.slack.com/archives/CAC6K3SLU/p1671050331922549?thread_ts=1670267924.080049&cid=CAC6K3SLU

Initial results are looking promising https://docs.google.com/document/d/1j0PTj5NpygdWARfcNy1ynwljQNrrX7rse86V4MjcdTU/edit#heading=h.yphl6brmvr69

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2179#-:-NJl1kLZAmfdewHKx4Bt:b-lwwk4w)* status: 1 of 8 files reviewed, 1 unresolved discussion (waiting on @msbutler)

**Comment by msbutler** [comment]
@itsbilal @sumeerbhola now that the multi level mechanics in #2129 have merged, I think this is ready for a review. We can also wait to review until I run the diff through @jbowens workload replayer. Let me know!

**Review by sumeerbhola — COMMENTED** [review]
This is a small PR so the code part of the review will be very quick.
- The experiments are the significant piece. So I'd like to wait on those results.
- To understand and iterate on the behavior of the heuristic, both for these experiments and even after we enable it in production, we will want additional metrics and log statements. I hacked some things (linked to from the doc mentioned above) which can use some clean up or rethink. It's fine to iterate over these as we start looking at experimental results, because it is hard to fully predict what is useful.
- Related to the above, I was reading the https://www.vldb.org/pvldb/vol15/p3071-dayan.pdf paper. We have known that what we are doing with multi-level compactions is akin to the multi file merge that systems that don't have multiple files per level do. The paper calls this "preemptive merge", and does talk about how it is better for write amp (at least for uniformly distributed writes -- no surprises there). We are generalizing "preemptive merge" for setups with multiple files per level. Given that:
  - I think we should log the score of all the levels picked in the multi-level compaction when it gets picked. For these high write throughput experiments, it is likely that the score is always close to 1, but (a) it will help understand the behavior once it is in production, (b) when new levels open up, even for these high throughput experiments, their score can initially be low.
  - In "debug mode", it would be good to understand how the write amp compares with kMinOverlappingRatio (see the comment in `compactionPickerByScore.pickFile`). That is, if we have decided to do a multi-level compaction involving Li, Li+1, Li+2, it would be good to know the ratio of file sizes picked in Li+2/Li+1, and what would the ratio have been if the file picked in Li+1 was picked using the kMinOverlappingRatio.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2179#-:-NMdMJhc4ZeS994ehEYG:bjdy6fm)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @itsbilal and @msbutler)

File: compaction_picker.go [hunk 1]
```diff
@@ -1586,6 +1586,42 @@ func (nml NoMultiLevel) pick(
 	return pc
 }
 
+func (pc *pickedCompaction) predictedWriteAmp() float64 {
```
Comment (msbutler) [inline compaction_picker.go:1589]: when evaluating the size of the files to compact,  I wonder if I should I create a new `files.SizeSumWithoutTombstones()` which subtracts each file's `file.stats.PointDeletionsBytesEstimate` or `file.stats.RangeDeletionsBytesEstimate`?

**Review by msbutler — COMMENTED** [review]
I'm working on a patch to gather these additional metrics, though I've got a few questions on their intended use:
- multilevel compaction writes: I see you amended the metrics table to include, for a given level: 1) total bytes written via multilevel compaction; 2) bytes written via multilevel compaction from the input level; 3) bytes rewritten " "' from the output level. These metrics seem like a fine way to summarize the impact of ML compactions, unless you had a specific gripe with them? 
- logging score for each level involved in the compaction: is the idea here to understand if we're inadvertently compacting down to a level that might have a high score? i.e. the multilevel output level itself deserves a compaction, which we didn't account for when choosing an ML compaction.
- computing a "kMinOverlappingRatio counterfactual": to clarify, when you say "what would the ratio have been if the file picked in Li+1 was picked using the kMinOverlappingRatio", are you saying: "independent of the files picked by the multilevel compaction, at this pebble version, what would have been the OverlappingRatio if we began a single level compaction at Li+1", right?  Is the idea then to compare the multi level overlapping ratio, i.e. `sum(FilesPicked in Li+2) / sum(FilesPicked in Li+1 _and_ Li)` to this hypothetical L1 overlapping ratio described above? Specifically, we want to know: "given that we're doing a compaction from Li+1 to Li+2, is multilevel choosing files in Li+2, that lead to a relatively poor (higher) Overlapping Ratio?"

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2179#-:-NN3jTr43mMi_mmqegP2:b-jov1hr)* status: 0 of 8 files reviewed, 2 unresolved discussions (waiting on @itsbilal and @sumeerbhola)

___
*[`cmd/pebble/db.go` line 75 at r3](https://reviewable.io/reviews/cockroachdb/pebble/2179#-NJkRBVQ5TUzCD3G7jnb:-NJkYi09CPPKVuuF0oNs:bfyq5xc) ([raw file](https://github.com/cockroachdb/pebble/blob/0415eeb51face21ba51adb5c6dd1c81dd118d61b/cmd/pebble/db.go#L75)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Do you know how many multi-level compactions happened in your ycsb/f experiments, with this addPropensity=0?
</blockquote></details>

By number of workers:
- 64: 12k
- 96:   17k
- 128: 13k

**Review by sumeerbhola — COMMENTED** [review]
> 1) total bytes written via multilevel compaction; 2) bytes written via multilevel compaction from the input level; 3) bytes rewritten " "' from the output level. These metrics seem like a fine way to summarize the impact of ML compactions, unless you had a specific gripe with them?

I didn't have a gripe with them :). The motivation for 1) was to see how many bytes flowing into a level were due to multi-level compaction (we can compare that with all the bytesin to that level, which is all compactions) -- this will tell us how often (in a byte-weighted manner) we were running multi-level compactions. 2) and 3) are where a ratio of 3) to 2) gives us a sense on how much write amp this multi-level compaction incurs.

> logging score for each level involved in the compaction: is the idea here to understand if we're inadvertently compacting down to a level that might have a high score? i.e. the multilevel output level itself deserves a compaction, which we didn't account for when choosing an ML compaction.

That's one way to utilize it. The other thing we could understand is if the middle level is not very full: say we compacted, L4, L5, L6 in a multi-level compaction and we see that we are often doing this even if L5 has a score that is low.

> computing a "kMinOverlappingRatio counterfactual": to clarify, when you say "what would the ratio have been if the file picked in Li+1 was picked using the kMinOverlappingRatio", are you saying: "independent of the files picked by the multilevel compaction, at this pebble version, what would have been the OverlappingRatio if we began a single level compaction at Li+1", right?

Correct. It gives us the ability to compare how good or bad in terms of the Li+2/Li+1 ratio this multi-level compaction is.

> Is the idea then to compare the multi level overlapping ratio, i.e. sum(FilesPicked in Li+2) / sum(FilesPicked in Li+1 _and_ Li) to this hypothetical L1 overlapping ratio described above? Specifically, we want to know: "given that we're doing a compaction from Li+1 to Li+2, is multilevel choosing files in Li+2, that lead to a relatively poor (higher) Overlapping Ratio?"

Yes, something like that. If we start seeing bad choices based on these log statements we can then use those to refine the heuristic.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2179#-:-NNJwTZJ1Kc9ZZpDDZjI:b-rmbswh)* status: 0 of 8 files reviewed, 2 unresolved discussions (waiting on @itsbilal)

**Comment by msbutler** [comment]
@sumeerbhola I've added two commits to the PR which I plan to benchmark.
1. additional logging as described above.
2. Previously we didn't allow multilevel compactions to start at L0, for no real reason. This commit allows this.

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2179)* status: 0 of 8 files reviewed, 2 unresolved discussions (waiting on @itsbilal and @sumeerbhola)

___
*[`compaction_picker.go` line 1589 at r5](https://reviewable.io/reviews/cockroachdb/pebble/2179#-NImFcHbw3j6VlEBOLlR-r5-1589:-NNca2zb-Jj4au8rzbGi:b-zfud4t) ([raw file](https://github.com/cockroachdb/pebble/blob/c8f5af62d05835d6353ad14556a799ef44344fe0/compaction_picker.go#L1589)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

SizeSum is jut summing the file sizes and doesn't do any compensation. So I don't quite understand the question. The compensation code is trying to pick the file in a level, that will be compacted down, hence we try to use the tombstone knowledge to pick the most beneficial file. It isn't clear to me whether that is relevant here, where we have already picked the file, and we are trying to decide between a two-level and three-level compaction. And we don't actually know which level those keys are in that will get deleted  (they could be in Li+3), so it isn't necessarily fair to use that to compute the write amp.
There is also the compensation that happens in https://github.com/cockroachdb/pebble/blob/a0da8a967640361124f7221230933ebe6960497c/compaction_picker.go#L1098-L1108 which one could use to decrease the size of Li+2 in this calculation.

I would also be ok with not doing any compensation for now.
</blockquote></details>

Apologies, my question lacked context. What I'm wondering is: when I estimate writeAmp here, pre compaction, I wonder if we ought to take into account tombstones in the participating inputFiles. 

But I think you answered my question. We don't currently have a good way to estimate tombstone size by level, given that "we don't actually know which level those keys are in that will get deleted (they could be in Li+3)"

**Review by sumeerbhola — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2179)* status: 0 of 10 files reviewed, 2 unresolved discussions (waiting on @itsbilal and @msbutler)

___
*[`compaction_picker.go` line 1589 at r5](https://reviewable.io/reviews/cockroachdb/pebble/2179#-NImFcHbw3j6VlEBOLlR-r5-1589:-NNhFQuI8tOoojkeGXu-:b-8fx2u7) ([raw file](https://github.com/cockroachdb/pebble/blob/c8f5af62d05835d6353ad14556a799ef44344fe0/compaction_picker.go#L1589)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Apologies, my question lacked context. What I'm wondering is: when I estimate writeAmp here, pre compaction, I wonder if we ought to take into account tombstones in the participating inputFiles. 

But I think you answered my question. We don't currently have a good way to estimate tombstone size by level, given that "we don't actually know which level those keys are in that will get deleted (they could be in Li+3)"
</blockquote></details>

:+1:

**Pull request closed without merging**